tasks/main.yml

```diff
@@ -117,32 +123,41 @@
     - ansible_distribution in ['CentOS', 'RedHat']
     - ansible_distribution_version is version('9', '>=')
 
-- name: Join to a specific Domain Controller
-  shell: |
-    set -euo pipefail
-    echo {{ ad_integration_password | quote }} | realm join -U \
-      {{ ad_integration_user | quote }} --membership-software \
-      {{ ad_integration_membership_software | quote }} \
-      {{ ad_integration_join_to_dc | quote }}
+- name: Build Command - Join to a specific Domain Controller
+  set_fact:
+    join_command: |
+      set -euo pipefail
+      echo {{ ad_integration_password | quote }} | realm join -U \
+        {{ ad_integration_user | quote }} --membership-software \
+        {{ ad_integration_membership_software | quote }} \
+        {{ ad_integration_additional_parameters }} \
+        {{ ad_integration_join_to_dc | quote }}
   when:
     - ad_integration_join_to_dc is not none
-    - not ansible_check_mode
-  register: __realm_join_output
-  failed_when:
-    - __realm_join_output is failed
-    - not __realm_join_output.stderr is search("Already joined to this domain")
-  changed_when: not __realm_join_output.stderr is
-    search("Already joined to this domain")
 
-- name: Perform discovery-based realm join operation
-  shell: |
-    set -euo pipefail
-    echo {{ ad_integration_password | quote }} | realm join -U \
-      {{ ad_integration_user | quote }} --membership-software \
-      {{ ad_integration_membership_software | quote }} \
-      {{ ad_integration_realm | quote }}
+- name: Build Join Command - Perform discovery-based realm join operation
+  set_fact:
+    join_command: |
+      set -euo pipefail
+      echo {{ ad_integration_password | quote }} | realm join -U \
+        {{ ad_integration_user | quote }} --membership-software \
+        {{ ad_integration_membership_software | quote }} \
+        {{ ad_integration_additional_parameters }} \
+        {{ ad_integration_realm | quote }}
   when:
     - ad_integration_join_to_dc is none
+
+- name: Show the join command for debug
+  debug:
+    msg: "Would run: '{{ join_command }}'"
+  when:
+    - ad_integration_join_to_dc == __ad_integration_dummy_dc or ad_integration_realm == __ad_integration_dummy_realm
+    - ansible_check_mode
+
+- name: Run realm join command
+  shell: "{{ join_command }}"
+  when:
+    - ad_integration_join_to_dc != __ad_integration_dummy_dc and ad_integration_realm != __ad_integration_dummy_realm
```

Comment:
Out of curiosity..., what happens if ad_integration_join_to_dc == __ad_integration_dummy_dc or ad_integration_realm == __ad_integration_dummy_realm _and ansible_check_mode is false?

Reply:
It should not try to join the realm, as dummy means that we are in a system test environment without AD so the task should be skipped and debug message should be produced instead.

Reply:
Yes, I understood. I was wondering if the value is accidentally set by a customer with ansible_check_mode is false, does it fail with an appropriate error?

Reply:
Well, it is a special value just for system tests, the role would skip running the realm join and just show what it would run instead.
If customer sets it, then the packages for ad integration will be installed and config with a dummy/test values will be created.
The role execution passes as expected.